rlay-backend-neo4j/src/lib.rs

```diff
@@ -220,76 +220,170 @@ impl Neo4jBackend {
-                            cid, key, value
-                        );
-                relationships.push(rel_query);
-            };
-
-            for (key, value) in val {
+            #[derive(Serialize, Deserialize, Debug)]
+            struct RelationshipQueryPart {
+                cid: String,
+                kind_name: String
+            }
+
+            // fetch the parts of the payload that should become relationships
+            // aka. those that are not self-referencing CIDs
+            for (key, value) in val.iter() {
                 if key == "cid" || key == "type" {
                     continue;
                 }
                 if (kind_name == "DataPropertyAssertion"
                     || kind_name == "NegativeDataPropertyAssertion")
                     && key == "target"
                 {
-                    values.push(format!("n.{0} = {1}", key, value));
                     continue;
                 }
                 if kind_name == "Annotation" && key == "value" {
-                    values.push(format!("n.{0} = {1}", key, value));
                     continue;
                 }
                 if let Value::Array(array_val) = value {
-                    for relationship_value in array_val {
-                        add_relationship_value(cid.clone(), key, relationship_value);
+                    for array_val_cid in array_val {
+                        relationships.push(RelationshipQueryPart{
+                            cid: array_val_cid.as_str().unwrap().to_string(),
+                            kind_name: key.clone()
+                        });
                     }
                     continue;
                 }
                 if let Value::String(_) = value {
-                    add_relationship_value(cid.clone(), key, value);
+                    relationships.push(RelationshipQueryPart{
+                        cid: value.as_str().unwrap().to_string(),
+                        kind_name: key.clone()
+                    });
                 }
             }
         }
 
-        let mut statement_query = format!(
-            "MERGE (n:RlayEntity {{cid: \"{1}\"}}) SET n:{0}",
-            kind_name, cid
+        let statement_query_main = format!("
+            MERGE (n:RlayEntity {{cid: $entity.cid }})
+            SET (CASE WHEN $entity.type = 'Annotation' THEN n END).value = $entity.value
+            SET (CASE WHEN $entity.type = 'DataPropertyAssertion' THEN n END).target = $entity.target
+            SET (CASE WHEN $entity.type = 'NegativeDataPropertyAssertion' THEN n END).target = $entity.target
+            WITH n
+            CALL apoc.create.addLabels(n, [$entity.type]) YIELD node
+            UNWIND $relationships as relationship
+                MERGE (m:RlayEntity {{ cid: relationship.cid }})
+                WITH n, m, relationship
+                CALL apoc.merge.relationship(n, relationship.kind_name, {{}}, {{}}, m) YIELD rel
+            RETURN n.cid",
```

Comment:
you either want to `RETURN distinct $entity.cid` or just `RETURN count(*)` otherwise every statements returns as many rows as you have relationships.

Reply:
that is a really good point. will push a change for this